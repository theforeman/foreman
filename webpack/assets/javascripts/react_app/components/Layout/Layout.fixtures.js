const mockOnClick = jest.fn();

const PFitems = [
  {
    title: 'Monitor',
    initialActive: true,
    iconClass: 'fa fa-tachometer',
    subItems: subItemsA,
    href: '/a',
  },
  {
    title: 'Hosts',
    initialActive: false,
    iconClass: 'fa fa-server',
    subItems: subItemsB,
    href: '/b',
  },
  {
    title: 'Configure',
    initialActive: false,
    iconClass: 'fa fa-wrench',
    subItems: subItemsC,
    href: '/c',
  },
];

const subItemsA = [
  {
    title: 'Aa',
    isDivider: false,
    onClick: mockOnClick,
  },
  {
    title: 'Cc',
    isDivider: false,
    onClick: mockOnClick,
  },
];
const subItemsB = [
  {
    title: 'Dd',
    isDivider: false,
    onClick: mockOnClick,
  },
  {
    title: 'Ff',
    isDivider: false,
    onClick: mockOnClick,
  },
];
const subItemsC = [
  {
    title: 'Gg',
    isDivider: false,
    onClick: mockOnClick,
  },
  {
    title: 'Ii',
    isDivider: false,
    onClick: mockOnClick,
  },
];

// Server Hash Data
const monitorChildren = [
  {
    type: 'item',
    name: 'Dashboard',
    url: '/',
  },
  {
    type: 'item',
    name: 'Facts',
    url: '/fact_values',
  },
];

const hostsChildren = [
  {
    type: 'item',
    name: 'All Hosts',
    url: '/hosts/new',
  },
  {
    type: 'item',
    name: 'Architectures',
    url: '/architectures',
  },
];

const userChildren = [
  {
    type: 'item',
    name: 'Environments',
    url: '/environments',
  },
  {
    type: 'item',
    name: 'Architectures',
    url: '/architectures',
  },
];
const infrastructureChildren = [
  {
    type: 'item',
    name: 'Domains',
    url: '/domains',
  },
  {
    type: 'item',
    name: 'Realms',
    url: '/realms',
  },
];

const hashItemsA = [
  {
    type: 'sub_menu',
    name: 'Monitor',
    icon: 'fa fa-tachometer',
    children: monitorChildren,
    active: false,
  },
  {
    type: 'sub_menu',
    name: 'Hosts',
    icon: 'fa fa-server',
    children: hostsChildren,
    active: false,
  },
];

const hashItemsB = [
  {
    type: 'sub_menu',
    name: 'User',
    icon: 'fa fa-wrench',
    children: userChildren,
    active: false,
  },
  {
    type: 'sub_menu',
    name: 'Infrastructure',
    icon: 'pficon pficon-network',
    children: infrastructureChildren,
    active: false,
  },
];

export const serverItems = [...hashItemsA, ...hashItemsB];

const logo =
  '/assets/header_logo-c9614c16f2ee399ae9cb7f36ec94b9a26bf8cf9eabaa7fe6099bf80d1f7940db.svg';
const user = {
  current_user: {
    user: {
      id: 4,
      login: 'admin',
      firstname: 'Admin',
      lastname: 'User',
    },
  },
  user_dropdown: [
    {
      type: 'sub_menu',
      name: 'User',
      icon: 'fa fa-user',
      children: subItemsA,
    },
  ],
};

const organizations = {
  current_org: 'org1',
  available_organizations: [
    { id: 1, title: 'org1', href: '/organizations/1-org1/select' },
    { id: 2, title: 'org2', href: '/organizations/2-org2/select' },
  ],
};

const locations = {
  current_location: 'london',
  available_locations: [
    { id: 1, title: 'yaml', href: '/locations/1-yaml/select' },
    { id: 2, title: 'london', href: '/locations/2-london/select' },
    { id: 3, title: 'norway', href: '/locations/3-norway/select' },
  ],
};

const serverUser = {
  current_user: {
    user: {
      firstname: 'G',
      lastname: 'L',
    },
  },
  user_dropdown: [
    {
      children: [
        {
          type: 'item',
          url: '/',
          name: 'My Account',
        },
        { type: 'divider' },
      ],
    },
  ],
};

export const layoutMock = {
  items: PFitems,
  data: {
    menu: [...hashItemsA, ...hashItemsB],
    locations,
    orgs: organizations,
    root: '/',
    logo,
    notification_url: '/notification_recipients',
    taxonomies: { locations: true, organizations: true },
    user,
  },
};

export const noItemsMock = {
  ...layoutMock,
  items: [],
};

export const hasTaxonomiesMock = {
  ...layoutMock,
  currentLocation: 'london',
  currentOrganization: 'org1',
};

export const userDropdownProps = {
  user: serverUser,
  notification_url: '/',
};
