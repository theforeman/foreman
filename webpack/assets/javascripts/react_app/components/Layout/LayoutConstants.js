export const LAYOUT_SHOW_LOADING = 'LAYOUT_SHOW_LOADING';
export const LAYOUT_HIDE_LOADING = 'LAYOUT_HIDE_LOADING';
export const LAYOUT_CHANGE_LOCATION = 'LAYOUT_CHANGE_LOCATION';
export const LAYOUT_CHANGE_ORG = 'LAYOUT_CHANGE_ORG';
export const LAYOUT_UPDATE_ITEMS = 'LAYOUT_UPDATE_ITEMS';
