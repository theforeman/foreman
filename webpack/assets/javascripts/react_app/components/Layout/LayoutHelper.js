import { isEmpty } from 'lodash';
import {
  changeOrganization,
  changeLocation,
} from '../../../foreman_navigation';
import { translate as __ } from '../../common/I18n';

export const getCurrentPath = () => window.location.pathname;

export const getActive = (data, path) => {
  let activeItem = '';
  data.forEach(item => {
    item.children.forEach(child => {
      if (child.url === path) activeItem = item.name;
    });
  });
  return { title: activeItem };
};

export const handleMenuClick = (primary, activeMenu, changeActive) => {
  if (primary.title !== activeMenu) changeActive(primary);
};

export const combineMenuItems = data => {
  const items = [];

  data.menu.forEach(item => {
    const translatedChildren = item.children.map(child => ({
      ...child,
      name: isEmpty(child.name) ? child.name : __(child.name),
    }));

    const translatedItem = {
      ...item,
      name: __(item.name),
      children: translatedChildren,
      // Hiding user if not on Mobile view
      className: item.name === 'User' ? 'visible-xs-block' : '',
    };
    items.push(translatedItem);
  });

  if (data.taxonomies.organizations) {
    items.push(createOrgItem(data.orgs.available_organizations));
  }

  if (data.taxonomies.locations) {
    items.push(createLocationItem(data.locations.available_locations));
  }
  return items;
};

const createOrgItem = orgs => {
  const anyOrg = {
    name: __('Any Organization'),
    url: '/organizations/clear',
    onClick: () => {
      changeOrganization(__('Any Organization'));
    },
  };
  const childrenArray = [];
  childrenArray.push(anyOrg);

  orgs.forEach(org => {
    const childObject = {
      type: org.type,
      name: isEmpty(org.title) ? org.title : __(org.title),
      onClick: () => {
        changeOrganization(__(org.title));
      },
      url: org.href,
    };
    childrenArray.push(childObject);
  });

  const orgItem = {
    type: 'sub_menu',
    name: __('Organizations'),
    icon: 'fa fa-building',
    children: childrenArray,
    // Hiding Organizations if not on Mobile view
    className: 'visible-xs-block',
    active: false,
  };
  return orgItem;
};

const createLocationItem = locations => {
  const anyLoc = {
    name: __('Any Location'),
    url: '/locations/clear',
    onClick: () => {
      changeLocation(__('Any Location'));
    },
  };
  const childrenArray = [];
  childrenArray.push(anyLoc);

  locations.forEach(loc => {
    const childObject = {
      type: loc.type,
      name: isEmpty(loc.title) ? loc.title : __(loc.title),
      onClick: () => {
        changeLocation(__(loc.title));
      },
      url: loc.href,
    };
    childrenArray.push(childObject);
  });

  const locItem = {
    type: 'sub_menu',
    name: __('Locations'),
    icon: 'fa fa-globe',
    children: childrenArray,
    // Hiding Locations if not on Mobile view
    className: 'visible-xs-block',
    active: false,
  };
  return locItem;
};
