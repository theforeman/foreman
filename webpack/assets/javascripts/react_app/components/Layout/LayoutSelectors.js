import { createSelector } from 'reselect';
import { get } from 'lodash';

export const selectLayout = state => state.layout;

export const selectMenuItems = state => selectLayout(state).items;
export const selectActiveMenu = state => selectLayout(state).activeMenu;
export const selectIsLoading = state => selectLayout(state).isLoading;
export const selectCurrentLocation = state =>
  get(selectLayout(state), 'currentLocation.title');
export const selectCurrentOrganization = state =>
  get(selectLayout(state), 'currentOrganization.title');

export const patternflyMenuItemsSelector = createSelector(
  selectMenuItems,
  selectCurrentLocation,
  selectCurrentOrganization,
  (items, currentLocation, currentOrganization) =>
    patternflyItems(items, currentLocation, currentOrganization)
);

const patternflyItems = (data, currentLocation, currentOrganization) => {
  if (data.length === 0) return [];
  const items = [];

  data.forEach(item => {
    const childrenArray = [];
    item.children.forEach(child => {
      const childObject = {
        title: child.name,
        isDivider: child.type === 'divider' && !!child.name,
        className:
          child.name === currentLocation || child.name === currentOrganization
            ? 'mobile-active'
            : '',
        href: child.url ? child.url : '#',
        preventHref: false,
        onClick: child.onClick ? () => child.onClick() : null,
      };
      childrenArray.push(childObject);
    });
    const itemObject = {
      title: item.name,
      initialActive: item.active,
      iconClass: item.icon,
      subItems: childrenArray,
      className: item.className,
    };
    items.push(itemObject);
  });
  return items;
};
