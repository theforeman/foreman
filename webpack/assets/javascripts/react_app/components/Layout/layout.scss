@import '../../../../../../app/assets/stylesheets/colors.scss';

.nav-pf-secondary-nav {
  background-color: $nav-pf-vertical-active-bg-color !important;

  .mobile-active {
    // Mark active Location/Org in Mobile view
    a {
      .list-group-item-value {
        background-color: $nav-pf-vertical-secondary-active-bg-color !important;
      }
    }
  }

  .list-group-item {
    a {
      background-color: $nav-pf-vertical-active-bg-color !important;

      &:hover {
        .list-group-item-value {
          background-color: $nav-pf-vertical-secondary-active-bg-color !important;
        }
      }
    }

    span {
      color: $nav-pf-vertical-active-icon-color !important;
    }
  }
}

.nav-pf-vertical {
  background-color: $nav-pf-vertical-secondary-active-bg-color !important;

  .list-group {
    .list-group-item {
      a {
        span {
          color: $nav-pf-vertical-icon-color;
        }

        &:hover {
          background-color: $nav-pf-vertical-active-bg-color;

          span {
            color: $nav-pf-vertical-active-icon-color;
          }
        }

        &::after {
          color: $nav-pf-vertical-icon-color;
        }
      }

      &:hover {
        a {
          background-color: $nav-pf-vertical-active-bg-color !important;

          span {
            color: $nav-pf-vertical-active-icon-color;
          }
        }
      }

      &.active {
        a {
          background-color: $nav-pf-vertical-active-bg-color;

          span {
            color: $nav-pf-vertical-active-icon-color;
          }
        }
      }
    }
  }
}

#account_menu {
  outline: none;
}

#vertical-spinner {
  margin-top: 15px;
  margin-left: 5px;
}

.navbar-header {
  width: 221.22px;
}
