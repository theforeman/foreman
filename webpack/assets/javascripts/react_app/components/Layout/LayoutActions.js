import { combineMenuItems } from './LayoutHelper';
import { selectMenuItems } from './LayoutSelectors';
import {
  LAYOUT_SHOW_LOADING,
  LAYOUT_HIDE_LOADING,
  LAYOUT_UPDATE_ITEMS,
  LAYOUT_CHANGE_LOCATION,
  LAYOUT_CHANGE_ORG,
} from './LayoutConstants';

export const showLoading = () => ({
  type: LAYOUT_SHOW_LOADING,
});

export const hideLoading = () => ({
  type: LAYOUT_HIDE_LOADING,
});

export const changeActiveMenu = primary => (dispatch, getState) => {
  const items = selectMenuItems(getState()).map(item =>
    primary.title === item.name
      ? { ...item, active: true }
      : { ...item, active: false }
  );

  dispatch({
    type: LAYOUT_UPDATE_ITEMS,
    payload: {
      items,
      activeMenu: primary.title,
    },
  });
};

export const fetchMenuItems = data => dispatch => {
  const items = combineMenuItems(data);
  dispatch({
    type: LAYOUT_UPDATE_ITEMS,
    payload: {
      items,
    },
  });
};

export const changeOrganization = org => dispatch => {
  dispatch({
    type: LAYOUT_CHANGE_ORG,
    payload: { org },
  });
};

export const changeLocation = location => dispatch => {
  dispatch({
    type: LAYOUT_CHANGE_LOCATION,
    payload: {
      location,
    },
  });
};
