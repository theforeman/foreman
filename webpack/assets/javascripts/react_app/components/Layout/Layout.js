import React from 'react';
import PropTypes from 'prop-types';

import { VerticalNav } from 'patternfly-react';
import { noop } from '../../common/helpers';

import { getActive, getCurrentPath, handleMenuClick } from './LayoutHelper';
import TaxonomySwitcher from './components/TaxonomySwitcher';
import UserDropdowns from './components/UserDropdowns';
import './layout.scss';

class Layout extends React.Component {
  componentDidMount() {
    const {
      items,
      data,
      fetchMenuItems,
      changeLocation,
      currentLocation,
      changeOrganization,
      currentOrganization,
      changeActiveMenu,
      activeMenu,
    } = this.props;
    if (items.length === 0) fetchMenuItems(data);

    const activeURLMenu = getActive(data.menu, getCurrentPath());
    if (activeMenu !== activeURLMenu.title) {
      changeActiveMenu(activeURLMenu);
    }

    if (
      data.taxonomies.locations &&
      !!data.locations.current_location &&
      currentLocation !== data.locations.current_location
    ) {
      const initialLocTitle = data.locations.current_location;
      const initialLocId = data.locations.available_locations.find(
        loc => loc.title === initialLocTitle
      ).id;
      changeLocation({ title: initialLocTitle, id: initialLocId });
    }

    if (
      data.taxonomies.organizations &&
      !!data.orgs.current_org &&
      currentOrganization !== data.orgs.current_org
    ) {
      const initialOrgTitle = data.orgs.current_org;
      const initialOrgId = data.orgs.available_organizations.find(
        org => org.title === initialOrgTitle
      ).id;
      changeOrganization({ title: initialOrgTitle, id: initialOrgId });
    }
  }

  render() {
    const {
      items,
      data,
      isLoading,
      changeActiveMenu,
      changeOrganization,
      changeLocation,
      currentOrganization,
      currentLocation,
      activeMenu,
    } = this.props;

    return (
      <VerticalNav
        hoverDelay={100}
        items={items}
        onItemClick={primary =>
          handleMenuClick(primary, activeMenu, changeActiveMenu)
        }
        activePath={`/${activeMenu}/`}
        {...this.props}
      >
        <VerticalNav.Masthead>
          <VerticalNav.Brand
            title={data.brand}
            iconImg={data.logo}
            href={data.root}
          />
          <TaxonomySwitcher
            taxonomiesBool={data.taxonomies}
            currentLocation={currentLocation}
            locations={
              data.taxonomies.locations
                ? data.locations.available_locations
                : []
            }
            onLocationClick={changeLocation}
            currentOrganization={currentOrganization}
            organizations={
              data.taxonomies.organizations
                ? data.orgs.available_organizations
                : []
            }
            onOrgClick={changeOrganization}
            isLoading={isLoading}
          />
          <UserDropdowns
            notificationUrl={data.notification_url}
            user={data.user}
            changeActiveMenu={changeActiveMenu}
          />
        </VerticalNav.Masthead>
      </VerticalNav>
    );
  }
}

Layout.propTypes = {
  currentOrganization: PropTypes.string,
  currentLocation: PropTypes.string,
  isLoading: PropTypes.bool,
  activeMenu: PropTypes.string,
  fetchMenuItems: PropTypes.func,
  changeActiveMenu: PropTypes.func,
  changeOrganization: PropTypes.func,
  changeLocation: PropTypes.func,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string.isRequired,
      className: PropTypes.string,
      iconClass: PropTypes.string.isRequired,
      initialActive: PropTypes.bool,
      subItems: PropTypes.arrayOf(
        PropTypes.shape({
          title: PropTypes.string,
          isDivider: PropTypes.bool,
          className: PropTypes.string,
          href: PropTypes.string.isRequired,
        })
      ),
    })
  ),
  data: PropTypes.shape({
    menu: PropTypes.arrayOf(
      PropTypes.shape({
        type: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        icon: PropTypes.string.isRequired,
        children: PropTypes.any,
      })
    ),
    locations: PropTypes.shape({
      current_location: PropTypes.string,
      available_locations: PropTypes.arrayOf(
        PropTypes.shape({
          href: PropTypes.string.isRequired,
          id: PropTypes.number.isRequired,
          title: PropTypes.string,
        })
      ),
    }),
    orgs: PropTypes.shape({
      current_organization: PropTypes.string,
      available_organizations: PropTypes.arrayOf(
        PropTypes.shape({
          href: PropTypes.string.isRequired,
          id: PropTypes.number.isRequired,
          title: PropTypes.string,
        })
      ),
    }),
    root: PropTypes.string.isRequired,
    logo: PropTypes.string.isRequired,
    notification_url: PropTypes.string.isRequired,
    taxonomies: PropTypes.shape({
      locations: PropTypes.bool.isRequired,
      organizations: PropTypes.bool.isRequired,
    }),
    user: PropTypes.shape({
      current_user: PropTypes.object.isRequired,
      user_dropdown: PropTypes.arrayOf(
        PropTypes.shape({
          children: PropTypes.any,
          icon: PropTypes.string.isRequired,
          name: PropTypes.string.isRequired,
          type: PropTypes.string.isRequired,
        })
      ),
    }),
  }),
};

Layout.defaultProps = {
  items: [],
  data: {},
  currentOrganization: 'Any Organization',
  currentLocation: 'Any Location',
  isLoading: false,
  activeMenu: '',
  fetchMenuItems: noop,
  changeActiveMenu: noop,
  changeOrganization: noop,
  changeLocation: noop,
};

export default Layout;
