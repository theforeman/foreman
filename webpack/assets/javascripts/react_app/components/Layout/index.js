import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './LayoutActions';
import reducer from './LayoutReducer';
import { selectLayout, patternflyMenuItemsSelector } from './LayoutSelectors';

import Layout from './Layout';

// map state to props
const mapStateToProps = (state) => {
  const layoutState = selectLayout(state);

  return {
    items: patternflyMenuItemsSelector(state),
    isLoading: layoutState.isLoading,
    activeMenu: layoutState.activeMenu,
    currentOrganization: layoutState.currentOrganization,
    currentLocation: layoutState.currentLocation,
  };
};

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { layout: reducer };

// export connected component
export default connect(mapStateToProps, mapDispatchToProps)(Layout);
