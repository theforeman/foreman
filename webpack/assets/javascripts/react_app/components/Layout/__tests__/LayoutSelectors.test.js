import { patternflyMenuItemsSelector } from '../LayoutSelectors';
import { layoutMock } from '../Layout.fixtures';

const state = {
  layout: {
    items: layoutMock.data.menu,
    activeMenu: 'Hosts',
  },
};
describe('Layout Selectors', () => {
  it('should return PF-React Compatible items', () => {
    const output = patternflyMenuItemsSelector(state);

    expect(output).toMatchSnapshot();
  });
  it('should return empty array', () => {
    const emptyState = {
      layout: {
        items: [],
      },
    };
    const output = patternflyMenuItemsSelector(emptyState);

    expect(output).toMatchSnapshot();
  });
});
