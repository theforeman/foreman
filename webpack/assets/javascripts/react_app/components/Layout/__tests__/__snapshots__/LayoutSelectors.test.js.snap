// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Selectors should return PF-React Compatible items 1`] = `
Array [
  Object {
    "className": undefined,
    "iconClass": "fa fa-tachometer",
    "initialActive": false,
    "subItems": Array [
      Object {
        "className": "",
        "href": "/",
        "isDivider": false,
        "onClick": null,
        "preventHref": false,
        "title": "Dashboard",
      },
      Object {
        "className": "",
        "href": "/fact_values",
        "isDivider": false,
        "onClick": null,
        "preventHref": false,
        "title": "Facts",
      },
    ],
    "title": "Monitor",
  },
  Object {
    "className": undefined,
    "iconClass": "fa fa-server",
    "initialActive": false,
    "subItems": Array [
      Object {
        "className": "",
        "href": "/hosts/new",
        "isDivider": false,
        "onClick": null,
        "preventHref": false,
        "title": "All Hosts",
      },
      Object {
        "className": "",
        "href": "/architectures",
        "isDivider": false,
        "onClick": null,
        "preventHref": false,
        "title": "Architectures",
      },
    ],
    "title": "Hosts",
  },
  Object {
    "className": undefined,
    "iconClass": "fa fa-wrench",
    "initialActive": false,
    "subItems": Array [
      Object {
        "className": "",
        "href": "/environments",
        "isDivider": false,
        "onClick": null,
        "preventHref": false,
        "title": "Environments",
      },
      Object {
        "className": "",
        "href": "/architectures",
        "isDivider": false,
        "onClick": null,
        "preventHref": false,
        "title": "Architectures",
      },
    ],
    "title": "User",
  },
  Object {
    "className": undefined,
    "iconClass": "pficon pficon-network",
    "initialActive": false,
    "subItems": Array [
      Object {
        "className": "",
        "href": "/domains",
        "isDivider": false,
        "onClick": null,
        "preventHref": false,
        "title": "Domains",
      },
      Object {
        "className": "",
        "href": "/realms",
        "isDivider": false,
        "onClick": null,
        "preventHref": false,
        "title": "Realms",
      },
    ],
    "title": "Infrastructure",
  },
]
`;

exports[`Layout Selectors should return empty array 1`] = `Array []`;
